### PR TITLE
[WIP] plotting atf_results

### LINKED ISSUE #11: atf presenter with pyplot
details see https://github.com/floweisshardt/atf/issues/108
fixes https://github.com/floweisshardt/atf/issues/108

### PR DESCRIPTION
**Pull request by floweisshardt**
fixes https://github.com/mojin-robotics/atf/issues/11

### CODE REVIEW

**Comment by floweisshardt** [comment]
I get the following errors when using the `plot.py` script (running `rosrun atf_plotter plot.py info-structure` works):

```
rosrun atf_plotter plot.py plot-metric -m tf_length_translation -tb testblock_small -t ts0_c0_r0_e0_s0_0 /tmp/atf_test_app_navigation/results_txt/atf_result.txt 
loading file... DONE (took 108.325s)
single_test.repetition_number 0
Traceback (most recent call last):
  File "/home/fmw/git/atf/src/atf/atf_plotter/scripts/plot.py", line 615, in <module>
    test=argparse_result.test
  File "/home/fmw/git/atf/src/atf/atf_plotter/scripts/plot.py", line 307, in plot_data_and_groundtruth_for_given_metric_testblock_test
    test_block = single_test[testblock]
  File "/home/fmw/git/atf/src/atf/atf_plotter/scripts/plot.py", line 81, in __getitem__
    raise KeyError
KeyError
```

and 
```
rosrun atf_plotter plot.py plot-metric -m tf_length_translation -tb testblock_small -t ts0_c0_r0_e0_s0_0 /tmp/atf_test/results_txt/atf_result.txt 
loading file... DONE (took 0.152s)
single_test.repetition_number 0
Traceback (most recent call last):
  File "/home/fmw/git/atf/src/atf/atf_plotter/scripts/plot.py", line 615, in <module>
    test=argparse_result.test
  File "/home/fmw/git/atf/src/atf/atf_plotter/scripts/plot.py", line 324, in plot_data_and_groundtruth_for_given_metric_testblock_test
    rects = ax.bar(idx - width / 2.0, metric.data, width, label=label)
  File "/home/fmw/git/atf/src/atf/atf_plotter/scripts/plot.py", line 117, in data
    return self._data_dict.get('data', None).get('data', None)
AttributeError: 'str' object has no attribute 'get'
```

File: atf_plotter/scripts/plot.py [hunk 1]
```diff
@@ -0,0 +1,761 @@
+#!/usr/bin/env python2
```
Comment (lindemeier) [inline atf_plotter/scripts/plot.py:1]: @fmessmer Should we use 
```
#!/usr/bin/env python
```
or explictitly
```
#!/usr/bin/env python3
```
due to the change to python3

**Review by lindemeier — COMMENTED** [review]
Cannot say much about the implementation. Looks good. 

More documentation or implementation detail comments would be appreciated.

File: atf_plotter/scripts/plot.py (hunk 1 above)
Comment (fmessmer) [reply atf_plotter/scripts/plot.py:1]: hmm, we haven't really focused on that aspect during the `pylint` session
personally, I'd vote for version-independent `#!/usr/bin/env python`...but we also have quite a number of python files using `#!/usr/bin/env python2`...

@LoyVanBeek
what's you opinion on this?

once decided, maybe we should be consequent about this and change it everywhere....
I added a reference to this comment to https://github.com/mojin-robotics/cob4/issues/1406

Comment (LoyVanBeek) [reply atf_plotter/scripts/plot.py:1]: It should be just `python` AFAIK and let the distro decide eventually

**Comment by LoyVanBeek** [comment]
Some test results:

```bash
lvb@lvb-xps-13 $ rosrun atf_plotter plot.py info-structure /tmp/mojin_atf_manipulation/results_txt/atf_result.txt
\loading file... DONE (took 0.003s)
ts0_c0_r0_e0_s0_0
    testblock_all_boxes
        1x user_result 
    testblock_expected_boxes
        1x user_result 
    testblock_manipulation
        1x user_result 
        1x time 
ts0_c1_r0_e0_s0_0
    testblock_all_boxes
        1x user_result 
    testblock_expected_boxes
        1x user_result 
    testblock_manipulation
        1x user_result 
        1x time 
[~/git/mojin_ws/src/mojin_atf/mojin_atf_navigation] (fix/use_DataStamped $=)
lvb@lvb-xps-13 $ rosrun atf_plotter plot.py plot-metric -t ts0_c0_r0_e0_s0_0 -tb testblock_manipulation -m time /tmp/mojin_atf_navigation/results_txt/atf_result.txt
loading file... DONE (took 0.017s)
single_test.repetition_number 0
Traceback (most recent call last):
  File "/home/lvb/git/mojin_ws/src/atf/atf_plotter/scripts/plot.py", line 734, in <module>
    test=argparse_result.test
  File "/home/lvb/git/mojin_ws/src/atf/atf_plotter/scripts/plot.py", line 302, in plot_data_and_groundtruth_for_given_metric_testblock_test
    test_block = single_test[testblock]
  File "/home/lvb/git/mojin_ws/src/atf/atf_plotter/scripts/plot.py", line 80, in __getitem__
    raise KeyError('%s not found in atf result' % key)
KeyError: 'testblock_manipulation not found in atf result'
```

So, `info-structure` reports there has to be a `testblock_manipulation` with a `time` metric (which is correct given the config). But I get a `KeyError: 'testblock_manipulation not found in atf result'`

Same for `rosrun atf_plotter plot.py plot-metric -t ts0_c0_r0_e0_s0_0 -tb testblock_forward -m tf_length_translation /tmp/mojin_atf_manipulation/results_txt/atf_result.txt`
 which has structure 
```bash
rosrun atf_plotter plot.py info-structure /tmp/mojin_atf_navigation/results_txt/atf_result.txt
loading file... DONE (took 0.016s)
ts0_c0_r0_e0_s0_0
    testblock_forward
        1x interface 
        2x tf_length_translation 
        1x user_result 
        1x time 
    testblock_all
        1x interface 
        2x tf_length_translation 
        1x user_result 
        1x time 
...
```

File: atf_plotter/scripts/plot.py (hunk 1 above)
Comment (Acuda) [reply atf_plotter/scripts/plot.py:1]: still prefer the aspect "explicit is better than implicit" from python design goals and would let it be python2 until script supports python3.

**Comment by LoyVanBeek** [comment]
I've been running some tests with data gathered by @floweisshardt as I only had data in `/tmp` and rebooted for the first time in weeks :roll_eyes: 

- [x] OK `rosrun atf_plotter plot.py -h`
- [x] OK `rosrun atf_plotter plot.py info-structure /tmp/mojin_atf_manipulation/results_txt/atf_result.txt`
- [x] OK `rosrun atf_plotter plot.py info-structure /home/lvb/ATF/atf_test_app_navigation/results_txt/atf_result.txt`
- [x] OK `rosrun atf_plotter plot.py plot-metric -t ts0_c0_r0_e0_s0_0 -tb testblock_all -m time /home/lvb/ATF/atf_test_app_navigation/results_txt/atf_result.txt`
- [x] OK `rosrun atf_plotter plot.py plot-metric -t ts0_c0_r0_e0_s0_0 -tb testblock_roundtrip -m tf_length_translation /home/lvb/ATF/atf_test_app_navigation/results_txt/atf_result.txt`
- [x] OK `rosrun atf_plotter plot.py plot-metric -t ts0_c0_r0_e0_s0_0 -tb testblock_small -m tf_length_translation /home/lvb/ATF/atf_test/results_txt/atf_result.txt`
- [x] OK `rosrun atf_plotter plot.py plot-b -t ts0_c0_r0_e0_s0_0 -tb testblock_small /home/lvb/ATF/atf_test/results_txt/atf_result.txt`
- [x] OK `rosrun atf_plotter plot.py plot-b -t ts0_c0_r0_e0_s0_0 -tb testblock_large /home/lvb/ATF/atf_test/results_txt/atf_result.txt`
- [x] OK `rosrun atf_plotter plot.py plot-c /home/lvb/ATF/atf_test/results_txt/atf_result.txt`
- [x] work but unclear what a 'test ident' is: `rosrun atf_plotter plot.py plot-d -ti=ts0_c0_r0_e0_s0 /home/lvb/ATF/atf_test/results_txt/atf_result.txt`

**Comment by Acuda** [comment]
> ... work but unclear what a 'test ident' is ...

`--help` is your friend, e.g. `rosrun atf_plotter plot.py plot-d --help` leads to:

```
usage: plot.py plot-d [-h] --testident TEST_CASE_IDENT

optional arguments:
  -h, --help            show this help message and exit
  --testident TEST_CASE_IDENT, -ti TEST_CASE_IDENT
                        like test name without repetition, e.g.
                        ts0_c0_r0_e0_s0
```

and the explanation "like test name without repetition, e.g. ts0_c0_r0_e0_s0"

still unclear "what a 'test ident' is"? if so, then i am happy if you could suggest a better help-text for the argument.

**Comment by LoyVanBeek** [comment]
The help text helped a bit, but I just don't know what 'ident' means. 
Is 'testcase enumeration' a correct interpretation?

**Comment by floweisshardt** [comment]
superceeded by https://github.com/floweisshardt/atf/pull/140

**Pull request closed without merging**